keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ResourceEvent.kt

```diff
@@ -367,23 +374,37 @@ data class ResourceCheckUnresolvable(
     )
 }
 
+enum class ResourceCheckErrorOrigin {
+  USER, SYSTEM;
+  companion object {
+    fun fromException(e: SpinnakerException) =
+      when (e) {
+        is UserException -> USER
+        is SystemException -> SYSTEM
+        else -> throw SystemException("All keel exceptions should inherit from UserException or SystemException. This is a bug.")
+      }
+  }
+}
+
 data class ResourceCheckError(
   override val kind: ResourceKind,
   override val id: String,
   override val application: String,
   override val timestamp: Instant,
-  val exceptionType: Class<out Throwable>,
-  val exceptionMessage: String?
-) : ResourceCheckResult() {
+  val exceptionType: Class<out SpinnakerException>,
+  val exceptionMessage: String?,
+  val origin: ResourceCheckErrorOrigin
+) : ResourceCheckResult(message = exceptionMessage) {
   @JsonIgnore
   override val state = Error
 
-  constructor(resource: Resource<*>, exception: Throwable, clock: Clock = Companion.clock) : this(
+  constructor(resource: Resource<*>, exception: SpinnakerException, clock: Clock = Companion.clock) : this(
     resource.kind,
     resource.id,
     resource.application,
     clock.instant(),
     exception.javaClass,
-    exception.message
+    exception.message,
+    ResourceCheckErrorOrigin.fromException(exception)
```

Comment:
This line will actually `throw` if the exception is not a `SystemException` or `UserException` with a proper error message. We'll only catch it a runtime, but at least we'll know if we messed up the types.